src/events.rs

```diff
@@ -71,20 +72,38 @@ impl std::fmt::Debug for RawEvent {
     }
 }
 
+/// Functions that take an input stream, consume an object, and output the serialized bytes.
+trait TypeSegmenterFn: Fn(&mut &[u8], &mut Vec<u8>) -> Result<(), Error> + Send {}
+impl<T> TypeSegmenterFn for T where
+    T: Fn(&mut &[u8], &mut Vec<u8>) -> Result<(), Error> + Send
+{
+}
+
```

Comment:
Wondering whether something like this would work:

```
trait TypeSegmenter<T: Codec> {
    fn segment(input: &mut &[u8], output: &mut Vec<u8>) -> Result<(), Error> {
        T::decode(input).map_err(Error::from)?.encode_to(output);
        Ok(())
    }
}

impl<T> TypeSegmenter<T> for T where T: Codec {}
```

Reply:
Nice suggestion! Implemented it as a trait, and use `Box<dyn TypeSegmenter>` to replace the closure